### ISSUE TITLE
Inconsistent indentation check results

### ISSUE THREAD

**Issue by morozov** [squizlabs/PHP_CodeSniffer#903]
The following is a simplified example of a legacy file which I work with and try to validate the changes with PHP_CodeSniffer:
```php
<?php

function foo()
{
·····if (true) {
·········echo 1;
·····}
·····if (true) {
·········echo 2;
·····}
}
```

The indentation inside `function` is 5 spaces, the indentation inside `if`-statements if 4 spaces. Here's the results which PHP_CodeSniffer produces for this file when checking against PSR2:
```
↪  scripts/phpcs --version
PHP_CodeSniffer version 2.6.0 (stable) by Squiz (http://www.squiz.net)

↪  scripts/phpcs ~/test.php --standard=PSR2

FILE: /home/morozov/test.php
----------------------------------------------------------------------
FOUND 5 ERRORS AFFECTING 5 LINES
----------------------------------------------------------------------
  5 | ERROR | [x] Line indented incorrectly; expected 4 spaces, found
    |       |     5
  7 | ERROR | [x] Line indented incorrectly; expected 8 spaces, found
    |       |     5
  8 | ERROR | [x] Line indented incorrectly; expected 8 spaces, found
    |       |     5
  9 | ERROR | [x] Line indented incorrectly; expected at least 12
    |       |     spaces, found 9
 10 | ERROR | [x] Line indented incorrectly; expected 8 spaces, found
    |       |     5
```

As you can see, it complains about line 9 (`echo 2`) but doesn't complain about line 6 (`echo 1`).

As far as I understand from reading the source code (`$adjustments` in `Generic_Sniffs_WhiteSpace_ScopeIndentSniff`), the offset inside blocks is counted individually, so both lines 6 and 9 should be considered indented correctly.

It's a bug and what is the expected result?

**Comment by gsherwood** [comment]
We must be looking at different code. When I test with this content:
```php
<?php

function foo()
{
     if (true) {
         echo 1;
     }
     if (true) {
         echo 2;
     }
}
```
Using the latest code from master, I get this output for the PSR2 standard:
```
FILE: temp.php
------------------------------------------------------------------------
FOUND 4 ERRORS AFFECTING 4 LINES
------------------------------------------------------------------------
  5 | ERROR | [x] Line indented incorrectly; expected 4 spaces, found 5
  7 | ERROR | [x] Line indented incorrectly; expected 4 spaces, found 5
  8 | ERROR | [x] Line indented incorrectly; expected 4 spaces, found 5
 10 | ERROR | [x] Line indented incorrectly; expected 4 spaces, found 5
------------------------------------------------------------------------
PHPCBF CAN FIX THE 4 MARKED SNIFF VIOLATIONS AUTOMATICALLY
------------------------------------------------------------------------
```
So I'd first recommend ensuring you have the latest code. If you do, and you get the same result, run this command to get a bit of extra debug output from the scope indent sniff:
`scripts/phpcs ~/test.php --standard=PSR2 --runtime-set scope_indent_debug 1`

The output I get from that is:
```
Start with token 0 on line 1 with indent 0
Open scope (T_FUNCTION) on line 4
	=> indent set to 4 by token 8 (T_OPEN_CURLY_BRACKET)
[Line 5] Line indented incorrectly; expected 4 spaces, found 5
	=> Add adjustment of -1 for token 11 (T_IF) on line 5
Open scope (T_IF) on line 5
	=> indent set to 8 by token 17 (T_OPEN_CURLY_BRACKET)
Indent adjusted to 8 for T_ECHO on line 6
	=> Add adjustment of -1 for token 20 (T_ECHO) on line 6
Close scope (T_IF) on line 7
	=> indent set to 4 by token 26 (T_CLOSE_CURLY_BRACKET)
[Line 7] Line indented incorrectly; expected 4 spaces, found 5
	=> Add adjustment of -1 for token 26 (T_CLOSE_CURLY_BRACKET) on line 7
[Line 8] Line indented incorrectly; expected 4 spaces, found 5
	=> Add adjustment of -1 for token 29 (T_IF) on line 8
Open scope (T_IF) on line 8
	=> indent set to 8 by token 35 (T_OPEN_CURLY_BRACKET)
Indent adjusted to 8 for T_ECHO on line 9
	=> Add adjustment of -1 for token 38 (T_ECHO) on line 9
Close scope (T_IF) on line 10
	=> indent set to 4 by token 44 (T_CLOSE_CURLY_BRACKET)
[Line 10] Line indented incorrectly; expected 4 spaces, found 5
	=> Add adjustment of -1 for token 44 (T_CLOSE_CURLY_BRACKET) on line 10
Close scope (T_FUNCTION) on line 11
	=> indent set to 0 by token 46 (T_CLOSE_CURLY_BRACKET)
```

**Comment by gsherwood** [comment]
Probably worth pointing out that I get the same result using version 2.5.1 as well, including the same debug output. I'm not aware of any commits in master that would have caused a change like this.

**Comment by morozov** [comment]
@gsherwood I'm able to reproduce the above issue on any version starting 2.3.0 (didn't try the earlier).

It looks insane, but with debug enabled, the issue is not reproducible:
```
↪  git rev-parse HEAD
be0a7044ec8f664c07f117e86633250122af6dbf

↪  scripts/phpcs ~/test.php --standard=PSR2 --runtime-set scope_indent_debug 1
Start with token 0 on line 1 with indent 0
Open scope (T_FUNCTION) on line 4
	=> indent set to 4 by token 8 (T_OPEN_CURLY_BRACKET)
[Line 5] Line indented incorrectly; expected 4 spaces, found 5
	=> Add adjustment of -1 for token 11 (T_IF) on line 5
Open scope (T_IF) on line 5
	=> indent set to 8 by token 17 (T_OPEN_CURLY_BRACKET)
Indent adjusted to 8 for T_ECHO on line 6
	=> Add adjustment of -1 for token 20 (T_ECHO) on line 6
Close scope (T_IF) on line 7
	=> indent set to 4 by token 26 (T_CLOSE_CURLY_BRACKET)
[Line 7] Line indented incorrectly; expected 4 spaces, found 5
	=> Add adjustment of -1 for token 26 (T_CLOSE_CURLY_BRACKET) on line 7
[Line 8] Line indented incorrectly; expected 4 spaces, found 5
	=> Add adjustment of -1 for token 29 (T_IF) on line 8
Open scope (T_IF) on line 8
	=> indent set to 8 by token 35 (T_OPEN_CURLY_BRACKET)
Indent adjusted to 8 for T_ECHO on line 9
	=> Add adjustment of -1 for token 38 (T_ECHO) on line 9
Close scope (T_IF) on line 10
	=> indent set to 4 by token 44 (T_CLOSE_CURLY_BRACKET)
[Line 10] Line indented incorrectly; expected 4 spaces, found 5
	=> Add adjustment of -1 for token 44 (T_CLOSE_CURLY_BRACKET) on line 10
Close scope (T_FUNCTION) on line 11
	=> indent set to 0 by token 46 (T_CLOSE_CURLY_BRACKET)

FILE: /home/morozov/test.php
----------------------------------------------------------------------
FOUND 4 ERRORS AFFECTING 4 LINES
----------------------------------------------------------------------
  5 | ERROR | [x] Line indented incorrectly; expected 4 spaces, found
    |       |     5
  7 | ERROR | [x] Line indented incorrectly; expected 4 spaces, found
    |       |     5
  8 | ERROR | [x] Line indented incorrectly; expected 4 spaces, found
    |       |     5
 10 | ERROR | [x] Line indented incorrectly; expected 4 spaces, found
    |       |     5
----------------------------------------------------------------------
PHPCBF CAN FIX THE 4 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------

Time: 16ms; Memory: 4Mb
```

While it's still reproducible without debug.

**Comment by morozov** [comment]
It's not reproducible with debug because of `if ($fix === true || $this->_debug === true)`. If I remove the debug part, the issue becomes reproducible. Not sure if it makes sense, though.

**Comment by gsherwood** [comment]
Ah, you've figured it out.

With debug enabled, the sniff applies those automated corrections, just like it does when fixing is enabled (during a phpcbf run). But without debug mode on, you don't get the same corrections because the fixer doesn't run at all (it would take more time for each run).

So what you're seeing is a side effect of that.

Here is the debug output without adjustment applied:
```
Start with token 0 on line 1 with indent 0
Open scope (T_FUNCTION) on line 4
	=> indent set to 4 by token 8 (T_OPEN_CURLY_BRACKET)
[Line 5] Line indented incorrectly; expected 4 spaces, found 5
Open scope (T_IF) on line 5
	=> indent set to 8 by token 17 (T_OPEN_CURLY_BRACKET)
Close scope (T_IF) on line 7
	=> indent set to 8 by token 26 (T_CLOSE_CURLY_BRACKET)
[Line 7] Line indented incorrectly; expected 8 spaces, found 5
[Line 8] Line indented incorrectly; expected 8 spaces, found 5
Open scope (T_IF) on line 8
	=> indent set to 12 by token 35 (T_OPEN_CURLY_BRACKET)
[Line 9] Line indented incorrectly; expected at least 12 spaces, found 9
Close scope (T_IF) on line 10
	=> indent set to 8 by token 44 (T_CLOSE_CURLY_BRACKET)
[Line 10] Line indented incorrectly; expected 8 spaces, found 5
Close scope (T_FUNCTION) on line 11
	=> indent set to 0 by token 46 (T_CLOSE_CURLY_BRACKET)
```
The error message `Line indented incorrectly; expected at least 12 spaces, found 9` should have been the give away, but I missed it.

Because no adjustments have been made, the first `echo` is looking for min indent of 8, which it finds. By the time checking gets to the second `echo`, the sniff has tried to get the indent to be divisible by 4, and set the `if` keyword to an indent of 8 spaces, which then asks the content of the `if` to be indented at least 12 spaces.

What you're seeing is just a result of the indents being in between two points and the fact that one error can cause additional ones.

Doing the extra checks to make the error messages 100% correct would be more time consuming (just like with the debug flag on) and slow down the check of every file. So it's probably not worth doing.

Does that at least explain what is happening here?

**Comment by morozov** [comment]
> Because no adjustments have been made, the first echo is looking for min indent of 8, which it finds. By the time checking gets to the second `echo`, the sniff has tried to get the indent to be divisible by 4, and set the `if` keyword to an indent of 8 spaces, which then asks the content of the `if` to be indented at least 12 spaces.

Yeah, this is close to what I figured out. Thank you.

> What you're seeing is just a result of the indents being in between two points and the fact that one error can cause additional ones.

> Doing the extra checks to make the error messages 100% correct would be more time consuming (just like with the debug flag on) and slow down the check of every file. So it's probably not worth doing.

Could you please describe the case where the slowdown would be significant or at least visible? I've tried checking a 8K LOC file with ~2K violations, and it takes exactly the same time (4.60 .. 4.63 seconds) to check this file both with or without that `if`-statement.

If it's still a problem, is it possible to introduce a run-time option for that?

**Comment by morozov** [comment]
I'd even say it _should_ be a separate option. Otherwise, debug mode influencing the results is kind of confusing.

**Comment by gsherwood** [comment]
I don't want an option for this because nobody would ever know it is there. So if I'm going to change the behaviour, I'll do it for everyone.

I was thinking more about checking entire repos, not so much single files, like you do with the Drupal and WordPress coding standards. But there might be a way to skip the fixing and assume it worked, because no other fixers would be running at the time.

I'll give that change a go and see if it works. If it does, it wont have any noticeable impact on performance because the fixer itself wouldn't actually be involved.

Thanks for prompting me to take another look. I was ready to close it :)

**Comment by gsherwood** [comment]
This is possible to do without reducing performance, but it radically changes the errors returned. When an entire block is indented too far, you'll now get an error for every single line inside that block instead of just getting errors for the start and end lines. Indenting the entire block will make them all go away, of course.

Not sure if this is a good thing or a bad thing, but it's a pretty big change. The next version is 2.6.0 and it doesn't technically break anything, so I'm leaning towards including it.

**Comment by gsherwood** [comment]
I made the change anyway. If a lot of people complain, I'll revert it in a future version. But the errors are technically more correct.

**Comment by morozov** [comment]
Thank you for considering this.
